modules/gapi/include/opencv2/gapi/gcompiled_async.hpp

```diff
@@ -12,20 +12,25 @@
 #include <exception>        //for std::exception_ptr
 #include <functional>       //for std::function
 #include <opencv2/gapi/garg.hpp>
+#include <opencv2/gapi/own/exports.hpp>
 
 namespace cv {
     //fwd declaration
     class GCompiled;
 
 namespace gapi{
 namespace wip {
+    class GAsyncContext;
     //These functions asynchronously (i.e. probably on a separate thread of execution) call operator() member function of their first argument with copies of rest of arguments (except callback) passed in.
     //The difference between the function is the way to get the completion notification (via callback or a waiting on std::future object)
     //If exception is occurred during execution of apply it is transfered to the callback (via function parameter) or passed to future (and will be thrown on call to std::future::get)
     GAPI_EXPORTS void                async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs);
+    GAPI_EXPORTS void                async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs, GAsyncContext& ctx);
+
     GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs);
-} // namespace gapi
+    GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs, GAsyncContext& ctx);
```

Comment:
I thought the idea was to have a default context so that there's no interface duplication just because there's a context. Something similar to this:
```suggestion
    GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs, GAsyncContext& ctx = {});
```

If you don't like the idea of getting `std::atomic<bool>` for no reason, let's put `DummyContext` to the API and call it `DefaultContext`

Reply:
context is passed in by reference , so having a default value for it means (without adding extra complexity, such shared internal state to the `GAsyncContext` itself) having a __global__ context object visible to everyone, thus ability to cancel __all__ async operations without explicit context, which is, presumably,  a bad idea.

Reply:
Hmm, you're right. I wonder, however, if we're actually going to modify the state of the context within the call. Most likely, context object is read-only for us, thus we can use `const GAsyncContext& ctx` which can be initialized with a temporary. Am I wrong?

My concern is in having `std::atomic<bool>` for no reason in case of default context

Reply:
Long story short - for "default" context no atomics are ever created

Reply:
True

Consider the comment resolved